unic/char/property/src/macros.rs

```diff
@@ -108,62 +136,122 @@ macro_rules! char_property {
-                    $(
-                        str if ::std::ascii::AsciiExt::eq_ignore_ascii_case(str, stringify!($abbr))
-                            => Ok($name::$variant),
-                        str if ::std::ascii::AsciiExt::eq_ignore_ascii_case(str, stringify!($long))
-                            => Ok($name::$variant),
-                    )*
-                    _ => Err(()),
-                }
-            }
+        char_property! {
+            __impl FromStr for $name;
+            $( $abbr => $name::$variant;
+               $long => $name::$variant; )*
         }
 
-        impl $crate::CharProperty for $name {
-            fn prop_abbr_name() -> &'static str { $abbr_name }
-            fn prop_long_name() -> &'static str { $long_name }
-            fn prop_human_name() -> &'static str { $human_name }
-        }
+        char_property! { __impl CharProperty for $name; $abbr_name; $long_name; $human_name; }
+        char_property! { __impl Display for $name by EnumeratedCharProperty }
 
         impl $crate::EnumeratedCharProperty for $name {
             fn all_values() -> &'static [$name] {
-                const VALUES: &[$name] = &[
-                    $($name::$variant,)+
-                ];
+                const VALUES: &[$name] = &[ $($name::$variant,)+ ];
                 VALUES
             }
-
             fn abbr_name(&self) -> &'static str {
-                match *self {
-                    $( $name::$variant => stringify!($abbr), )*
-                }
+                match *self { $( $name::$variant => stringify!($abbr), )* }
             }
-
             fn long_name(&self) -> &'static str {
-                match *self {
-                    $( $name::$variant => stringify!($long), )*
-                }
+                match *self { $( $name::$variant => stringify!($long), )* }
             }
-
             fn human_name(&self) -> &'static str {
-                match *self {
-                    $( $name::$variant => $human, )*
+                match *self { $( $name::$variant => $human, )* }
+            }
+        }
+    };
+
+    // == Binary Property == //
+
+    (
+        $(#[$name_meta:meta])* pub struct $name:ident(bool) {
+            abbr => $abbr_name:expr;
+            long => $long_name:expr;
+            human => $human_name:expr;
+        }
+
+        $(#[$data_mod_meta:meta])* mod $data_mod:ident = $data_path:expr;
+    ) => {
+        $(#[$name_meta])*
+        #[derive(Copy, Clone, Debug, Default, Eq, PartialEq, Hash)]
+        pub struct $name(bool);
+
+        $(#[$data_mod_meta])*
+        mod $data_mod {
+            use super::unic_utils::CharDataTable;
+            pub const TABLE: CharDataTable<()> = include!($data_path);
+        }
+
+        impl $name {
+            pub fn of(ch: char) -> Self { $name(data::TABLE.contains(ch)) }
+            pub fn bool(&self) -> bool { self.0 }
+        }
+
+        impl From<bool> for $name {
+            fn from(b: bool) -> Self { $name(b) }
```

Comment:
Actually, I don't think we want this one. We prefer to prevent creating values manually in general, right? I prefer we provide const values, like YES and NO, instead of conversion from boolean. What do you think?